### DISCUSSION TITLE
Where is the render() function of mitsuba integrator?

### DISCUSSION THREAD

**Discussion by ychanu** [mitsuba-renderer/mitsuba3#949] [Q&A] [answered]
Hi, I'm delving into the inner workings of the Mitsuba3 renderer. While adding debugging messages with Log(Debug) in the render() function of integrator.cpp, I observed that Mitsuba3 doesn't seem to utilize the render() functions from integrator.cpp.

For instance, even after removing debug messages like "Starting render job (??x??, ?? samples, ?? threads)", Mitsuba3 continued to display them.

Based on my debugging so far, when calling mitsuba.render(), it invokes the eval() method in util.py, which in turn calls self.integrator.render(). However, I'm struggling to identify where this render() function originates from. Is there anyone who can assist me with this?

**Comment by ziyi-zhang** [comment]
Hi,

It depends on the integrator.
If it's an integrator implemented in python (reparam, prb, etc), the entry point is [here](https://github.com/mitsuba-renderer/mitsuba3/blob/8033a807091f8315c5cef25f4f1a36a3766fb223/src/python/python/ad/integrators/common.py).

**Reply by ychanu** [reply]
Thank you for your comment.

Now,I'm working with the "path" integrator. 
The call sequence appears to be:  mitsuba.render() -> render() in util.py -> _RenderOp::eval() in util.py -> self.integrator.render() (???)

Do you happen to know the location of the last self.integrator.render() function?

I was expecting to find this in integrator.cpp, but it wasn't there during my debugging.

**Reply by merlinND** [reply, marked as answer]
Hi @ychanu,

`integrator.cpp` does implement the `render()` method:

https://github.com/mitsuba-renderer/mitsuba3/blob/8033a807091f8315c5cef25f4f1a36a3766fb223/src/render/integrator.cpp#L120-L126

**Reply by ychanu** [reply]
I initially expected that function to be called, but the path integrator still functions correctly even when I remove that code segment and return an empty tensor. 

It might be an issue on my end. I will study more by referring to the above answers.

Thank you for your thoughtful replies :)

**Reply by ychanu** [reply]
I solved this problem. It was a visual studio building problem. Thank you again!
